### PR TITLE
Add hierarchy to `Scope`

### LINKED ISSUE #42: Add Scope rules for hierarchy of org > campaign > project
Currently, `Scope` is just a triple of strings representing `org`, `campaign`, `project`, with no sense of hierarchy between these levels. This results in cumbersome access permissions for e.g. a user that should have access to an `org` called `openff`: they will require an explicit entry for each `campaign` and `project` represented in that `org` to have full access, with new entries added as new `campaign`s and `project`s are created.

To avoid this, we'd like for the hierarchy between these levels to be made explicit, and for glob-based wildcards to be allowed. So, we could have e.g. `*-*-*` (all scopes), `openff-*-*` (all `org` 'openff' scopes), `openff-rosemary-*` (all `org` 'openff', `campaign` 'rosemary' scopes), as allowed entries in a user's list of `Scope`s.

This hierarchy enforcement should disallow entries like `*-rosemary-*` (access to any `campaign` called 'rosemary', anywhere). In other words, one cannot limit scope at a level lower when given broadest scope at a level higher.

Ideally, `Scope` validity is enforced in the [Pydantic model itself](https://github.com/openforcefield/fah-alchemy/blob/0cdfa891cddbc55cf8d6daaaded719903dc89ad4/fah_alchemy/models.py#L6-L16) on creation.

### PR DESCRIPTION
**Pull request by hmacdope**
Partially fixes #42 

I have also added the `no-dash` validation for `ScopedKey`, as mentioned briefly in #42. Let me know if this is wrong and I can take it out. 

I also added a test for invalid scope creation with a dash in the name

I have also added support for "*" wildcards in addition to `None`

### CODE REVIEW

**Comment by hmacdope** [comment]
I should add that another possible way to do this is regex, but I'm not so good at that.

**Comment by hmacdope** [comment]
@dotsdl this is probably ready for a first look.

File: fah_alchemy/base/api.py [hunk 1]
```diff
@@ -79,7 +83,9 @@ def validate_scopes_query(
         ):
 
             # match (query_org == token_org) then (query_campaign == token_campaign) then (query_proj == token_proj)
-            if not (query_field == target_field or query_field is None):
+            if not (
+                query_field == target_field or query_field is None or query_field == "*"
```
Comment (dotsdl) [inline fah_alchemy/base/api.py:87]: We don't need to test for `'*'` here, as the field will actually be `None` if it's the wildcard value. We use `'*'` in the string representation, but in a `Scope` object this is actually represented as `None`.

This ends up having nice consequences when we use `Scope` objects to query Neo4j, since for a property dict keys with values of `None` end up not being included as query params, requiring us to do little to no processing in the query code.

Comment (dotsdl) [reply fah_alchemy/base/api.py:87]: I've changed `Scope._validate_component` to turn '*' into `None` when given as a component on `init`.

File: fah_alchemy/tests/unit/test_base_api.py [hunk 2]
```diff
@@ -15,6 +26,8 @@ def tokendata():
             "org1-campaignA-projectI",
             "org1-campaignB-projectI",
             "org1-campaignB-projectII",
+            "org2-*-*",
+            "org3-campaignA-*",
```
Comment (dotsdl) [inline fah_alchemy/tests/unit/test_base_api.py:30]: I added two additional `Scope`s to this user's authorized list. Can you:
1. fix the tests below that these additions broke, and
2. modify `base.api.validate_scopes` and `base.api.validate_scopes_query` such that they handle the presence of non-specific scopes such as these among a user's authorized set?
    - for `validate_scopes`, though `scope` may be specific, the `Scope`s in `token` won't necessarily be.
    - for `validate_scopes_query`, neither the `query_scope` or the `Scope`s in `token` will necessarily be specific.
    - for both these cases, it may make sense to add a `superset` method or similar to `Scope` that you can use for evaluating `Scope`s against each other. I've added a stub for this.

**Review by dotsdl — CHANGES_REQUESTED** [review]
Great work @hmacdope! I've added comments, along with some direction as to what I think we need yet. 

Now that we have hierarchical handling of components in `Scope`, we want to be able to add `Scope`s with wildcards to a user's authorized list. This means that any `Scope`-based queries or gets a user does with `FahAlchemyClient` must be evaluated against this list to ensure they *can* access the things they are asking for. This is tricky, but I think we're close.

File: fah_alchemy/base/api.py (hunk 1 above)
Comment (hmacdope) [reply fah_alchemy/base/api.py:87]: Okay makes sense to me to not to have "*" and instead cast to `None`

File: fah_alchemy/tests/unit/test_base_api.py (hunk 2 above)
Comment (hmacdope) [reply fah_alchemy/tests/unit/test_base_api.py:30]: @dotsdl I think this has the correct behaviour now. Feel free to take a look. :)

**Review by dotsdl — APPROVED** [review]
Great work @hmacdope! Feel free to merge!
